fvm/evm/emulator/emulator.go

```diff
@@ -0,0 +1,318 @@
+	return res, bl.commit(res.StateRootHash)
+}
+
+// RunTransaction runs an evm transaction
+func (bl *BlockView) RunTransaction(
+	tx *gethTypes.Transaction,
+) (*types.Result, error) {
+	var err error
+	proc, err := bl.newProcedure()
+	if err != nil {
+		return nil, err
+	}
+
+	msg, err := gethCore.TransactionToMessage(tx, GetSigner(bl.config), proc.config.BlockContext.BaseFee)
+	if err != nil {
+		// note that this is not a fatal error (e.g. due to bad signature)
+		// not a valid transaction
+		return nil, types.NewEVMValidationError(err)
+	}
+
+	// update tx context origin
+	proc.evm.TxContext.Origin = msg.From
+	res, err := proc.run(msg, tx.Type())
+	if err != nil {
+		return res, err
+	}
+
+	return res, bl.commit(res.StateRootHash)
+}
+
+func (bl *BlockView) newProcedure() (*procedure, error) {
+	execState, err := newState(bl.database)
+	if err != nil {
+		return nil, err
+	}
+	cfg := bl.config
+	return &procedure{
+		config: cfg,
+		evm: gethVM.NewEVM(
+			*cfg.BlockContext,
+			*cfg.TxContext,
+			execState,
+			cfg.ChainConfig,
+			cfg.EVMConfig,
+		),
+		state: execState,
+	}, nil
+}
+
+func (bl *BlockView) commit(rootHash gethCommon.Hash) error {
+	// commit atree changes back to the backend
+	err := bl.database.Commit(rootHash)
+	return handleCommitError(err)
+}
+
+type procedure struct {
+	config *Config
+	evm    *gethVM.EVM
+	state  *gethState.StateDB
+}
+
+// commit commits the changes to the state.
+func (proc *procedure) commit() (gethCommon.Hash, error) {
+	// commits the changes from the journal into the in memory trie.
+	// in the future if we want to move this to the block level we could use finalize
+	// to get the root hash
+	newRoot, err := proc.state.Commit(true)
+	if err != nil {
+		return gethTypes.EmptyRootHash, handleCommitError(err)
+	}
+
+	// flush the trie to the lower level db
+	// the reason we have to do this, is the original database
+	// is designed to keep changes in memory until the state.Commit
+	// is called then the changes moves into the trie, but the trie
+	// would stay in memory for faster transaction execution. you
+	// have to explicitly ask the trie to commit to the underlying storage
+	err = proc.state.Database().TrieDB().Commit(newRoot, false)
+	if err != nil {
+		return gethTypes.EmptyRootHash, handleCommitError(err)
+	}
+	return newRoot, nil
+}
+
+func handleCommitError(err error) error {
+	if err == nil {
+		return nil
+	}
+	// if known types (database errors) don't do anything and return
+	if types.IsAFatalError(err) || types.IsADatabaseError(err) {
+		return err
+	}
+
+	// else is a new fatal error
+	return types.NewFatalError(err)
+}
+
+func (proc *procedure) mintTo(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+		TxType:      types.DirectCallTxType,
+	}
+
+	// create account if not exist
+	if !proc.state.Exist(addr) {
+		proc.state.CreateAccount(addr)
```

Comment:
I don't think this is needed, add a balance call below will create a new object if it doesn't exist. I can understand tho if you wanted to make it super clear.

Reply:
I think add balance would create object not the account :-?